src/middlewared/middlewared/plugins/vm.py

```diff
@@ -515,6 +524,64 @@ def flags(self):
             self.logger.error("===> {0}".format(err))
             return False
 
+    @private
+    async def find_clone(self, name):
+        data = await self.middleware.call('vm.query', [], {'order_by': ['name']})
+        clone_index = 0
+        next_name = ""
+        for vm_name in data:
+            if name in vm_name['name'] and '_clone' in vm_name['name']:
+                name_index = int(vm_name['name'][-1])
+                next_name = vm_name['name'][:-1]
+                if name_index >= clone_index:
+                    clone_index = int(name_index) + 1
+
+        if next_name:
+            next_name = next_name + str(clone_index)
+        else:
+            next_name = name + '_clone' + str(clone_index)
+
+        return next_name
+
+    @accepts(Int('id'))
+    async def clone(self, id):
+        vm = await self._manager.clone(id)
+
+        if vm is None:
+            self.logger.error("Cannot clone a VM that does not exist.")
+            return False
+
+        origin_name = vm['name']
+        del vm['id']
+
+        vm['name'] = await self.find_clone(vm['name'])
+
+        for item in vm['devices']:
+            if item['dtype'] == 'NIC':
+                if 'mac' in item['attributes']:
+                    del item['attributes']['mac']
+            if item['dtype'] == 'VNC':
+                if 'vnc_port' in item['attributes']:
+                    del item['attributes']['vnc_port']
+            if item['dtype'] == 'DISK':
+                disk_src_path = '/'.join(item['attributes']['path'].split('/dev/zvol/')[-1:])
+                disk_snapshot_name = vm['name']
+                disk_snapshot_path = disk_src_path + '@' + disk_snapshot_name
+                clone_dst_path = disk_src_path + '_' + vm['name']
+
+                await self.middleware.call('zfs.snapshot.create', disk_src_path, disk_snapshot_name)
+                await self.middleware.call('zfs.snapshot.clone', disk_snapshot_path, clone_dst_path)
+
+                item['attributes']['path'] = '/dev/zvol/' + clone_dst_path
+            if item['dtype'] == 'RAW':
+                item['attributes']['path'] = ''
+                self.logger.warn("For RAW disk you need copy it manually inside your NAS.")
+
+        await self.do_create(vm)
```

Comment:
`do_create` is supposed to be a private method. since it implements the CRUDService type. I am planning on renaming and `do_*` to `_do_*` soon. Could you please rename it to just `create`?